scripts/04-find-patterns.py

```diff
@@ -20,208 +20,157 @@
-        + [x['code'] for x in config['metrics'] if x['code'] not in config['metrics_exclude']] \
-        + ['lines_' + x['code'] for x in config['patterns'] if x['code'] not in patterns_exclude] \
-        + ['filename']
-
-    with open(filename, 'w', newline='\n', encoding='utf-8') as csv_file:
-        writer = csv.DictWriter(
-            csv_file, delimiter=';',
-            quotechar='"',
-            quoting=csv.QUOTE_MINIMAL,
-            fieldnames=fields)
-        writer.writeheader()
-
-    pool = multiprocessing.Pool(multiprocessing.cpu_count())
-    log_func = partial(log_result, file_to_write=str(filename))
-    manager = Manager()
-    exceptions = manager.dict()
-    func = partial(execute_python_code_in_parallel_thread, exceptions)
-    with open(args.filename, 'r') as f:
-        file_names = [i for i in f.readlines()]
-    with open(filename, 'a', newline='\n', encoding='utf-8') as csv_file:
-        for result in pool.imap(func, file_names):
+        patterns_codes + \
+        metrics_codes + \
+        ["lines_" + code for code in patterns_codes] + \
+        ["filepath", "class_name", "component_index"]
+
+    return DictWriter(file, delimiter=";", quotechar='"', quoting=QUOTE_MINIMAL, fieldnames=fields)
+
+
+if __name__ == "__main__":
+    allowed_cores_qty = len(sched_getaffinity(0))
+    system_cores_qty = cpu_count()
+
+    parser = ArgumentParser(description="Creates dataset")
+    parser.add_argument("file", help="Path for file with a list of Java files.")
+
+    parser.add_argument(
+        "--jobs",
+        "-j",
+        type=int,
+        default=min(allowed_cores_qty, system_cores_qty - 1),
+        help="Number of processes to spawn. "
+        "By default one less than number of cores. "
+        "Be carefull to raise it above, machine may stop responding while creating dataset.",
+    )
+
+    parser.add_argument(
+        "--timeout",
+        "-t",
+        type=float,
+        default=60,
+        help="Maximum time (in seconds) for single file proccessing. Default is 60 seconds. "
+        "If 0, no timout is set.",
+    )
+
+    args = parser.parse_args()
+
+    if args.jobs >= system_cores_qty:
+        print(
+            f"WARNING: You have ordered to spawn {args.jobs} jobs, "
+            f"while system has only {system_cores_qty} cores. "
+            "Machine may badly respond, while calculating dataset.",
+            file=stderr,
+        )
+
+    if args.jobs > allowed_cores_qty:
+        print(
+            f"WARNING: You have ordered to spawn {args.jobs} jobs, "
+            f"while process only allowed to occupy {allowed_cores_qty} cores."
+            "You may have a slowdown due to large number of processes.",
+            file=stderr,
+        )
+
+    if args.timeout == 0:
+        args.timeout = None
+
+    default_dataset_directory = Path(".", "target", "04").absolute()
+    dataset_directory = getenv("TARGET_FOLDER", default=default_dataset_directory)
```

Comment:
make 
```
if target_folder:
    os.chdir(target_folder)
```, 

otherwise target folder won't work

Reply:
Changing directory might be quite dangerous. Anyway I use this directory in the following way `Path(dataset_directory, "04-find-patterns.csv")`, so dataset is created in right directory.